src/main/java/gregtech/api/capability/impl/miner/MinerLogic.java

```diff
@@ -0,0 +1,645 @@
+    }
+
+    /**
+     * Gets the blocks to mine
+     * @return a {@link LinkedList} of {@link BlockPos} for each ore to mine
+     */
+    private LinkedList<BlockPos> getBlocksToMine() {
+        LinkedList<BlockPos> blocks = new LinkedList<>();
+
+        // determine how many blocks to retrieve this time
+        double quotient = getQuotient(getMeanTickTime(metaTileEntity.getWorld()));
+        int calcAmount = quotient < 1 ? 1 : (int) (Math.min(quotient, Short.MAX_VALUE));
+        int calculated = 0;
+
+        // keep getting blocks until the target amount is reached
+        while (calculated < calcAmount) {
+            // moving down the y-axis
+            if (y.get() > 0) {
+                // moving across the z-axis
+                if (z.get() <= startZ.get() + currentRadius * 2) {
+                    // check every block along the x-axis
+                    if (x.get() <= startX.get() + currentRadius * 2) {
+                        BlockPos blockPos = new BlockPos(x.get(), y.get(), z.get());
+                        Block block = metaTileEntity.getWorld().getBlockState(blockPos).getBlock();
+                        if (block.blockHardness >= 0 && metaTileEntity.getWorld().getTileEntity(blockPos) == null && GTUtility.isOre(block)) {
+                            blocks.addLast(blockPos);
+                        }
+                        // move to the next x position
+                        x.incrementAndGet();
+                    } else {
+                        // reset x and move to the next z layer
+                        x.set(x.get() - currentRadius * 2);
+                        z.incrementAndGet();
+                    }
+                } else {
+                    // reset z and move to the next y layer
+                    z.set(z.get() - currentRadius * 2);
+                    y.decrementAndGet();
+                }
+            } else
+                return blocks;
+
+            // only count iterations where blocks were found
+            if (!blocks.isEmpty())
+                calculated++;
+        }
+        return blocks;
+    }
+
+    /**
+     * gets the quotient for determining the amount of blocks to mine
+     * @param base is a value used for calculation, intended to be the mean tick time of the world the miner is in
+     * @return the quotient
+     */
+    private static double getQuotient(double base) {
+        return DIVIDEND / Math.pow(base, POWER);
+    }
+
+    /**
+     * @param values to find the mean of
+     * @return the mean
+     */
+    private static long mean(long[] values) {
+        long sum = 0L;
+        for (long v : values)
+            sum += v;
+        return sum / values.length;
+    }
+
+    /**
+     *
+     * @param world the {@link World} the miner is in
+     * @return the mean tick time
+     */
+    private static double getMeanTickTime(World world) {
+        return mean(Objects.requireNonNull(world.getMinecraftServer()).tickTimeArray) * 1.0E-6D;
+    }
```

Comment:
These could be in `GTUtility` or another class in the `Util` package instead of here, for better references from other classes